Optimize finalized header storage

claravanstaden opened the pull request:
Overwrite the last stored finalized header if the header before that is more than SLOTS_PER_HISTORICAL_ROOT apart. This way, we only store the minimum finalized headers we need to prove any message in the past. This drastically reduces finalized header storage while widening the timeframe where we can prove messages. cc @alistair-singh thanks for the idea!

Resolves: [SNO-956](https://linear.app/snowfork/issue/SNO-956)
Snowbridge companion: https://github.com/Snowfork/snowbridge/pull/1175

yrong commented on bridges/snowbridge/pallets/ethereum-client/src/lib.rs:
```diff
@@ -497,10 +497,35 @@ pub mod pallet {
-				header_root,
-				CompactBeaconState { slot: header.slot, block_roots_root },
-			);
+			let index = FinalizedBeaconStateIndex::<T>::get();
+			// If the ringbuffer is at the first slot, then the previous head is the last slot
+			// because the ringbuffer wraps around
+			let second_last_index =
+				if index == 0 { MaxFinalizedHeadersToKeep::<T>::get() - 1 } else { index - 1 };
+
+			let prev_head_header_root = FinalizedBeaconStateMapping::<T>::get(second_last_index);
+			let prev_head_finalized_state = FinalizedBeaconState::<T>::get(prev_head_header_root)
+				.unwrap_or(CompactBeaconState::default());
+
+			// If the last finalized header and the one provided to be added to the store
+			// is less than a sync committee apart, overwrite the last finalized header since
+			// we only need one finalized update per SLOTS_PER_HISTORICAL_ROOT. If the slots
+			// are SLOTS_PER_HISTORICAL_ROOT apart, store the update in a new index in the
+			// ringbuffer.
```
I would suggest to move/refactor these logic to the impl of ringbuffer, a custom `RingBufferMapImplWithConditionalOverWrite` maybe. With a type param `CheckOverWriteTrait` as follows.

```
pub trait CheckOverWriteTrait<Value> {
	fn can_over_write(value: Value, prev_value: Value) -> bool;
}
```

yrong commented:
This PR reminds me of another improvement based on the previous analysis here
https://linear.app/snowfork/issue/SNO-805#comment-5c1ba857

To make finality update on demand, the previous beacon relayer degrades to only sync sync-committee update which will decrease the cost by 87%.

claravanstaden commented:
> This PR reminds me of another improvement maybe we can do.

> Based on the previous analysis here https://linear.app/snowfork/issue/SNO-805#comment-5c1ba857 It's possible to make finality update on demand, then the previous beacon relayer degrades to only sync sync-committee update which will decrease the cost by 87%. 

If I understand your comment in https://linear.app/snowfork/issue/SNO-805, we scan for messages on Ethereum, and only then submit a finalized header? That sounds great! I think that's a larger change though, and best suited for another PR, perhaps later. This is a quick win with minimal changes, which ensures we can prove messages waaay back.

yrong reviewed (approved):
Awesome!

vgeddes commented on bridges/snowbridge/pallets/ethereum-client/src/impls.rs:
```diff
@@ -159,3 +162,10 @@ impl<T: Config> Pallet<T> {
+
+impl<T: Config> CheckOverwrite<CompactBeaconState> for Pallet<T> {
+	fn can_overwrite(new_value: &CompactBeaconState, prev_value: &CompactBeaconState) -> bool {
+		let slots_difference = new_value.slot - prev_value.slot;
```
Need to handle underflow here.

claravanstaden commented:
> If we are only storing a finalized header every `SLOTS_PER_HISTORICAL_ROOT` (roughly 27 hours), then I am not sure I see a need for the ringbuffer at all. In which case we should consider removing it, and rework the code here.

Do you think so? I think its still worthwhile because it removes the need to think about pruning completely, which I think is useful. What do you think we gain from removing it?

claravanstaden commented:
> > > If we are only storing a finalized header every `SLOTS_PER_HISTORICAL_ROOT` (roughly 27 hours), then I am not sure I see a need for the ringbuffer at all. In which case we should consider removing it, and rework the code here.
> > 
> > 
> > Do you think so? I think its still worthwhile because it removes the need to think about pruning completely, which I think is useful. What do you think we gain from removing it?
> 
> * We gain robustness by removing especially complicated code paths
> * We reduce transaction costs as updating the ringbuffer is not cheap
> 
> Moreover, if `CompactBeaconState` is only 40 bytes, then a years worth of updates is only 14 KB which is insignificant.
> 
> Unrestricted state growth is the main reason for pruning, and your optimization reduces state growth to such low levels that I think we don't need a ring buffer

Fair enough. Let me remove it then. ✂️

claravanstaden commented:
@vgeddes removing the ringbuffer would requires some sort of storage migration, which complicates this work a bit but it is of course doable. Something that I am keeping in mind is that the relayer needs to be able to query the last finalized headers in order, so that if the beacon relayer restarts and the cache clears, the relayer can still fetch the last synced headers on-chain. So a simple map structure might not work, we need some kind of an index.

claravanstaden commented:
@vgeddes I added a quick idea here (not production grade or finished but just want to illustrate the idea): https://github.com/Snowfork/polkadot-sdk/pull/139/commits/8484f01badc5ec91318b98142fadbe380173db8f

Do you think just removing the ringbuffer is justified? We would still need different 3 storage items to keep track of the beacon headers anyway:
- A map of beacon header roots to compact beacon state (to lookup ancestry proof roots)
- A map of indexes to beacon header roots (to know which finalized header to overwrite)
- A counter to keep track of the last index imported (to know where in the map to overwrite)

vgeddes commented:
> * A map of beacon header roots to compact beacon state (to lookup ancestry proof roots)
> * A map of indexes to beacon header roots (to know which finalized header to overwrite)
> * A counter to keep track of the last index imported (to know where in the map to overwrite)

Yeah to clarify, when I meant remove the ring buffer, I also meant that we should remove the underlying storage items that are only required by it, namely `FinalizedBeaconStateMapping` and `FinalizedBeaconStateIndex`. We should be able to remove these since multi-block migrations should be [available](https://github.com/paritytech/polkadot-sdk/pull/1781) now. 

Then we are only left with `FinalizedBeaconState`, which we just fill up with headers and don't prune.

And `FinalizedBeaconState` would not need to be migratred

claravanstaden commented:
@vgeddes we need the ability to:

- know which header was added last and second last, which I don't think we can accomplish with just `FinalizedBeaconState` (correct me if I am wrong)
- lookup headers by header root

Afaik this would require us to keep `FinalizedBeaconStateMapping` and `FinalizedBeaconStateIndex`

claravanstaden commented:
@vgeddes I have updated the PR to have a smaller diff in the ringbuffer module and added more comments. Additionally, this supporting document describes the ringbuffer with examples: https://www.notion.so/snowfork/Ringbuffer-7330714108994775b219364a4ef06fd4.

vgeddes commented:
Thanks @claravanstaden, the notion doc helped a lot.

I realize that we still need the beacon relayer to submit beacon headers frequently so that new execution messages can be verified. While the optimization does reduce storage usage, I don't think it will actually reduce relaying costs. In fact costs will increase slightly as the optimization increases the weight of `submit`.

Stepping back, I wonder if there is an alternative that involves:
1. Enforcing a minimum and maximum number of slots between imported headers
2. Make the beacon relayer more intelligent so that it only submits beacon headers on demand (and at least every `SLOTS_PER_HISTORICAL_ROOT`).

This will accomplish two things:
* Reduce storage usage as relayers are not sending so frequently
* Reduce costs for relayers.

claravanstaden commented:
Closing in favour of [SNO-1015](https://linear.app/snowfork/issue/SNO-1015).
